atd-etl/cris_import/cris_import.py

```diff
@@ -225,45 +292,94 @@ def specify_extract_location():
     return zip_tmpdir
 
 
-def download_archives():
-    """
-    Connect to the SFTP endpoint which receives archives from CRIS and
-    download them into a temporary directory.
+def download_s3_archive():
+    session = boto3.Session(
+        aws_access_key_id=AWS_ACCESS_KEY_ID,
+        aws_secret_access_key=AWS_SECRET_ACCESS_KEY,
+    )
 
-    Returns path of temporary directory as a string
-    """
+    s3 = session.client("s3")
+    bucket = S3_EXTRACT_BUCKET
 
     with SshKeyTempDir() as key_directory:
         write_key_to_file(
-            key_directory + "/id_ed25519", SFTP_ENDPOINT_SSH_PRIVATE_KEY + "\n"
+            key_directory + "/id_ed25519", DB_BASTION_HOST_SSH_PRIVATE_KEY + "\n"
+        )
+        ssh_tunnel = SSHTunnelForwarder(
+            (DB_BASTION_HOST),
+            ssh_username=DB_BASTION_HOST_SSH_USERNAME,
+            ssh_private_key=f"{key_directory}/id_ed25519",
+            remote_bind_address=(DB_RDS_HOST, 5432),
         )
+        ssh_tunnel.start()
 
-        zip_tmpdir = tempfile.mkdtemp()
-        rsync = None
-        try:
-            rsync = sysrsync.run(
-                verbose=True,
-                options=["-a"],
-                source_ssh=SFTP_ENDPOINT,
-                source="/home/txdot/*zip",
-                sync_source_contents=False,
-                destination=zip_tmpdir,
-                private_key=key_directory + "/id_ed25519",
-                strict_host_key_checking=False,
+        pg = psycopg2.connect(
+            host="localhost",
+            port=ssh_tunnel.local_bind_port,
+            user=DB_USER,
+            password=DB_PASS,
+            dbname=DB_NAME,
+            sslmode=DB_SSL_REQUIREMENT,
+            sslrootcert="/root/rds-combined-ca-bundle.pem",
+        )
+
+        uploads_prefix = f"{DEPLOYMENT_ENVIRONMENT}/uploads/"
+        # Get list of all objects in the bucket with the specified prefix
+        objects = s3.list_objects(Bucket=bucket, Prefix=uploads_prefix)
+
+        cursor = pg.cursor()
+
+        for obj in objects["Contents"]:
+            full_object_path = obj["Key"]
+            object_name = os.path.basename(full_object_path)
+
+            # Skip the "folder" object
+            if full_object_path == uploads_prefix:
+                continue
+
+            # Extract the S3 prefix from the full object path
+            object_path = os.path.dirname(full_object_path)
+
+            # Extract the schema from the filename
+            schema = int(object_name.split("_")[1])
+
+            # Check if the object already exists in the database
+            cursor.execute(
+                "SELECT * FROM cris_import_log WHERE object_path = ? AND object_name = ?",
+                (object_path, object_name),
             )
-        except:
-            print("No files to copy..")
```

Comment:
@frankhereford what do you think about throwing an exception if there are no files to copy? i think this should be a pretty rare occurrence for our prod deployment, right?

Reply:
Given the recent events, I think this is an exceptional idea and I'll roll that into this PR. Great call.

Reply:
This will throw an exception now if it has no work to do -- thank you for this super good suggestion! https://github.com/cityofaustin/atd-vz-data/pull/1445/commits/df608325cbc7fb10108159713ad662397835bf3b

Reply:
all credit to @chiaberry